include/utility/array.h

```diff
@@ -0,0 +1,197 @@
+/*************************
+ * @file array.h
+ * @author zhuzhile08 (zhuzhile08@gmail.com)
+ * 
+ * @brief a basic array implementation
+ * 
+ * @date 2022-12-24
+ * 
+ * @copyright Copyright (c) 2022
+ *************************/
+#pragma once
+
+#include <utility>
+#include <algorithm>
+
+#define NODISCARD [[nodiscard]]
+#define DEPRECATED [[deprecated]]
```

Comment:
These defines may already exist

Reply:
Why are they being used at all?

Reply:
Normally, I'd have a central header with some macros and aliases like these, so I guess I could do that here too. I use these because they just look better and the standard uses macros like these too.